### PR TITLE
remote: add support for aliyun-oss

### PR DESCRIPTION
**Pull request by luolibin**
Can you help me test it and see where's the problem?

### CODE REVIEW

File: oss/oss.go [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+	meta, error2 := archiveBucket.GetObjectDetailedMeta(key)
+	if error2 != nil {
+		return error2
+	}
+	ContentLength, error3 := strconv.ParseInt(meta.Get("Content-Length"), 10, 64)
+	if error3 != nil {
+		return error3
+	}
+	o.size = ContentLength
+	return erro
```
Comment (ishuah) [inline oss/oss.go]: undefined `erro`

File: oss/oss.go [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+}
+
+// SetModTime sets the modification time of the local fs object
+func (o *Object) SetModTime(modTime time.Time) error {
+	err := o.metaData(o.fs)
+	if err != nil {
+		return err
+	}
+	o.meta = make(map[string]*string)
+	o.meta[metaMtime] = &(swift.TimeToFloatString(modTime))
```
Comment (ishuah) [inline oss/oss.go]: `cannot take the address of swift.TimeToFloatString(modTime)`

**Comment by ishuah** [comment]
Hi @luolibin 
Looks like some dependencies are not being met, specifically `github.com/aliyun/aliyun-oss-go-sdk/oss`. Run `dep ensure` to include this package in the list of vendors.

Aside from that I've highlighted a few syntax issues in your code.

This is just a cursory glance at your PR, hope it helps!

File: oss/oss.go [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+		return nil, err
+	}
+	for _, bucket := range listBucketsResult.Buckets {
+		d := fs.NewDir(bucket.Name, bucket.CreationDate)
+		entries = append(entries, d)
+	}
+	return entries, nil
+}
+
+// Returns objects and directories list ad fs.DirEntries
```
Comment (xerrni) [inline oss/oss.go]: change `ad` to `as` 
It was my misspelling.

File: oss/oss.go [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+// Returned entries will contain the list of all objects, however they can be placed out of order.
+func (f *Fs) List(dir string) (entries fs.DirEntries, err error) {
+	if f.bucket == "" {
+		return f.listBuckets(dir)
+	}
+	return f.listDir(dir)
+}
+
+// ListR recursively lists the objects and directories of the Fs.
```
Comment (xerrni) [inline oss/oss.go]: change `ListR` to `List`
It was my misspelling.

File: oss/oss.go [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+		return list.Add(entry)
+	})
+	if err != nil {
+		return err
+	}
+	return list.Flush()
+}
+
+// Function push input to the remote path.
+// sIf modTime or other parameters will be passed they will be pushed as well.
```
Comment (xerrni) [inline oss/oss.go]: change `sIf` to `If`

File: oss/oss.go [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+
+// Features returns the optional features of this Fs
+func (f *Fs) Features() *fs.Features {
+	return f.features
+}
+
+// Pattern to match a oss path
+var matcher = regexp.MustCompile(`^([^/]*)(.*)$`)
+
+// parseParse parses a oss 'url'
```
Comment (xerrni) [inline oss/oss.go]: change `parseParse` to `ossParsePath`

**Comment by ncw** [comment]
@luolibin what options to I need in the config?

I have

```
[TestOss]
type = oss
env_auth = false
access_key_id = XXX
secret_access_key = XXX
region = oss-eu-central-1
endpoint = oss-eu-central-1.aliyuncs.com
location_constraint = eu-central-1
acl = private
server_side_encryption = 
storage_class = 
```

But all I get back is `StatusCode=400, ErrorCode=InvalidArgument, ErrorMessage=Authorization header is invalid.`

Any ideas?

**Comment by ncw** [comment]
Figured it out!  Got the names of the keys wrong.

**Comment by ncw** [comment]
To debug this further this would be very useful: https://github.com/aliyun/aliyun-oss-go-sdk/issues/72 - I've made a PR for that.

Once that is in, you can add

```patch
--- a/oss/oss.go
+++ b/oss/oss.go
@@ -256,7 +256,7 @@ func ossParsePath(path string) (bucket, directory string, err error) {
 
 // ossConnection makes a connection to oss
 func ossConnection(endpoint string, accessID string, accessKey string) (*oss.Client, error) {
-	client, err := oss.New(endpoint, accessID, accessKey)
+	client, err := oss.New(endpoint, accessID, accessKey, oss.Transport(fs.Config.Transport()))
 	return client, err
 }
 ```

Then you can use `--dump-bodies` and `go test -v -verbose -dump-bodies`

The metadata doesn't appear to be working so the modification times aren't working. I haven't looked at the docs yet though.

It looks like you are doing very well (except for the metadata).

Do you want me to carry on from here, or do you want to have a go at fixing it up?

**Comment by luolibin** [comment]
@ncw I have met a  paroblem in metadata ,then I assign to o.size in the metadata method,and this problem seems to be solved.But then I still found metadata problems when unit testing.Can you help me find the problem if you are convenient.

**Comment by luolibin** [comment]
I'm refining my code.

**Comment by ncw** [comment]
I updated my PR https://github.com/aliyun/aliyun-oss-go-sdk/pull/76 to set an http.Client option.  Use it like this:

```patch
--- a/oss/oss.go
+++ b/oss/oss.go
@@ -256,7 +256,7 @@ func ossParsePath(path string) (bucket, directory string, err error) {
 
 // ossConnection makes a connection to oss
 func ossConnection(endpoint string, accessID string, accessKey string) (*oss.Client, error) {
-	client, err := oss.New(endpoint, accessID, accessKey)
+	client, err := oss.New(endpoint, accessID, accessKey, oss.HTTPClient(fs.Config.Client()))
 	return client, err
 }
 
```

If you use this patch with `go test -v -dump-bodies -verbose` you can see that the object metadata is never being sent in the PUT requests.  There are no `x-oss-meta-` headers anywhere.

**Comment by ncw** [comment]
Here is an outline of what you need to do to read and set the metadata (not very well tested!)

```diff
diff --git a/oss/oss.go b/oss/oss.go
index 4569300e..1282ad9f 100644
--- a/oss/oss.go
+++ b/oss/oss.go
@@ -3,6 +3,7 @@ package oss
 import (
 	"fmt"
 	"io"
+	"net/http"
 	"net/url"
 	"regexp"
 	"strconv"
@@ -615,7 +617,24 @@ func (o *Object) metaData(f *Fs) (err error) {
 		return error3
 	}
 	o.size = ContentLength
-	return erro
+	o.meta = make(map[string]*string)
+	// Read metadata map
+	for key, values := range meta {
+		for _, value := range values {
+			if strings.HasPrefix(value, oss.HTTPHeaderOssMetaPrefix) {
+				s := value[len(oss.HTTPHeaderOssMetaPrefix):]
+				o.meta[key] = &s
+			}
+		}
+	}
+	// Read last modified
+	lastModified := meta.Get("Last-Modified")
+	if t, err := time.Parse(http.TimeFormat, lastModified); err == nil {
+		o.lastModified = t
+	} else {
+		fs.Debugf(nil, "Couldn't parse lastmodified from %q: %v", lastModified, err)
+	}
+	return err
 }
 
 // ModTime returns the modification time of the object
@@ -647,7 +666,8 @@ func (o *Object) SetModTime(modTime time.Time) error {
 		return err
 	}
 	o.meta = make(map[string]*string)
-	o.meta[metaMtime] = &(swift.TimeToFloatString(modTime))
+	modTimeString := swift.TimeToFloatString(modTime)
+	o.meta[metaMtime] = &modTimeString
 	if o.size >= maxSizeForCopy {
 		fs.Debugf(o, "SetModTime is unsupported for objects bigger than %v bytes", fs.SizeSuffix(maxSizeForCopy))
 		return nil
@@ -695,7 +715,11 @@ func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOptio
 		return err
 	}
 	key := o.fs.root + o.remote
-	erro := bucket.PutObject(key, in, oss.ContentType(mimeType))
+	ossOptions := []oss.Option{oss.ContentType(mimeType)}
+	for k, v := range o.meta {
+		ossOptions = append(ossOptions, oss.Meta(k, *v))
+	}
+	erro := bucket.PutObject(key, in, ossOptions...)
 	if erro != nil {
 		return erro
 	}
```

**Comment by ncw** [comment]
...and one more thing...

Please don't add a submodule, [vendor the code in a separate commit](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#adding-a-dependency) - thanks!

**Comment by luolibin** [comment]
@ncw I will pay attention to the question you mentioned.There is a question，does Httpclient is necessary for rclone?

**Comment by ncw** [comment]
> There is a question，does Httpclient is necessary for rclone?

Some features of rclone won't work without it, eg --dump-bodies, --dump-headers, timeouts etc.  It isn't essential but it would be nice.

**Comment by luolibin** [comment]
It can not work with metadata when I do unit test...I suspect that the update method does not upload metadata.

**Comment by ncw** [comment]
> I suspect that the update method does not upload metadata,and modification time of the obejct which I copied have changed, it became the current time.

Check with `--dump-bodies` to see if the right things are being sent in the headers.

**Pull request closed without merging**